x-pack/legacy/plugins/lens/public/persistence/saved_object_store.ts

```diff
@@ -0,0 +1,75 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { SavedObjectAttributes } from 'target/types/server';
+
+export interface Document {
+  id?: string;
+  type?: string;
+  visualizationType: string | null;
+  datasourceType: string | null;
+  title: string;
+  state: {
+    datasource: unknown;
+    visualization: unknown;
+  };
+}
+
+interface SavedObectStore {
+  create: (type: string, object: SavedObjectAttributes) => Promise<{ id: string }>;
+  update: (type: string, id: string, object: SavedObjectAttributes) => Promise<{ id: string }>;
+  get: (
+    type: string,
+    id: string
+  ) => Promise<{ id: string; type: string; attributes: SavedObjectAttributes }>;
+}
+
+const DOC_TYPE = 'lens';
+
+export interface DocumentSaver {
+  save: (vis: Document) => Promise<{ id: string }>;
+}
+
+export interface DocumentLoader {
+  load: (id: string) => Promise<Document>;
+}
+
+export type SavedObjectStore = DocumentLoader & DocumentSaver;
+
+export class SavedObjectIndexStore implements SavedObjectStore {
```

Comment:
The methods here have no error handling, which feels like the place that error handling is most needed. This is directly calling the saved object service. If you look at my example from `loader.ts` you'll see something like this as the error condition:

```typescript
    .catch(err => {
      toastNotifications.addDanger('Failed to load index patterns');
    });
```